### PR TITLE
snapcraft.yaml: add support for building a snap package

### PR DESCRIPTION
**Pull request by 3v1n0**
The package is generated to be completely [confined](https://snapcraft.io/docs/reference/confinement) right now, and this might cause some minor annoyances, for example: it's not possible to access (sync) .dotted folders, the home is in the `~/snap/nextcloud-client`, password can't be saved to the system keyring.

If you prefer not have these limitations, and make this client to run without any sandbox, we can move to the `classic` confinement, and this will work as it used to be.

The build can also be easily integrated with travis-CI and you can upload the new revisions automatically to the 'edge' channel of the snap store.

If you need any help for this, I'm available.

### CODE REVIEW

**Comment by enoch85** [comment]
@kyrofa Is this correct? Can you review please?

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [inline linux/snap/snapcraft.yaml]: At least for the time being, Ubuntu Software shows this field as the title. I suggest capitalizing it: "Nextcloud Desktop Client."

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [inline linux/snap/snapcraft.yaml]: s/ownCloud/Nextcloud/, no?

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [inline linux/snap/snapcraft.yaml]: Personally, until LP: [#1653769](https://bugs.launchpad.net/snapd/+bug/1653769) is resolved, I'd prefer classic confinement. Password prompts at start-up make it too annoying for me to use.

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [inline linux/snap/snapcraft.yaml]: g++ is missing here.

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [inline linux/snap/snapcraft.yaml]: I suggest adding a .desktop file for this as well.

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [inline linux/snap/snapcraft.yaml]: This doesn't actually pick up the theme (which is why, as you may have noticed, everything still says ownCloud). `-DOEM_THEME_DIR=$SNAPCRAFT_PART_INSTALL/../src/nextcloudtheme` will make you happier. Note that you'll need to update other sections in the YAML (such as the .desktop file and LD_LIBRARY_PATHs) that are set for something related to ownCloud.

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [inline linux/snap/snapcraft.yaml]: Can you explain why `indicator-qt5` is required? I was able to get the icon working fine simply by setting TMPDIR.

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [inline linux/snap/snapcraft.yaml]: This wasn't picked up for me (i.e. doesn't show in the dash). Did you test this?

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [inline linux/snap/snapcraft.yaml:44]: Can you explain this part, please? Obviously snapd-xdg-open needs to run as an external "trusted helper" if you will (i.e. outside of confinement). You're just copying a piece of it so as to not use the one `xdg-utils`, it seems, but I was under the impression that nothing "special" was needed within a snap to use snapd-xdg-open as long as it was installed on the system. Is that not the case?

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [inline linux/snap/snapcraft.yaml]: I'm gonna go out on a limb and say you probably want something in `../nextcloudtheme/theme/colored/` :wink: .

**Review by kyrofa — CHANGES_REQUESTED** [review]
Thanks @3v1n0! Good work here, although I have a number of suggestions/comments/questions. And perhaps the kicker: this doesn't run at all for me, it just seems to hang. Not sure if that's because of the snap or because I'm trying to run master, though.

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [reply linux/snap/snapcraft.yaml]: Take a look at [this](https://github.com/kyrofa/nextcloud-client-snap) (and test with `sudo snap install --edge nextcloud-client-kyrofa` on amd64) to see what I did specifically. I'm curious what this part adds on top of that. As I mentioned, I couldn't get this to run, so I was unable to see for myself.

**Comment by enoch85** [comment]
Can you please fix the rest @3v1n0 Thanks! :)

File: linux/snap/snapcraft.yaml
Comment (3v1n0) [reply linux/snap/snapcraft.yaml]: Ok, agree... Still not sure if devmode is better though as an interim step to the final confined (if that is ever wanted).

File: linux/snap/snapcraft.yaml
Comment (3v1n0) [reply linux/snap/snapcraft.yaml]: That is an "hack" that I actually was using (although now you should now use `TMPDIR` set as `XDG_RUNTIME_DIR`, as that is guaranteed to be cleaned up), however when snaps are confined that's not enough because this also would try to connect to dbus to get if running under unity or not.

There's an SRU for appmenu-qt5 in progress which will make this optional, but in the mean time that is the cleanest solution for this problem (see trunk `appmenu-qt5` code).

File: linux/snap/snapcraft.yaml
Comment (3v1n0) [reply linux/snap/snapcraft.yaml:44]: This kind of `xdg-open` right now is needed in order to be able to call that trusted service.
At least, without it there's no way for me to open links (while `file://` protocol is not supported unfortunately) when confined.

**Comment by 3v1n0** [comment]
@kyrofa 

> And perhaps the kicker: this doesn't run at all for me

Ah, I've noticed that too when you only build it confined.
If you run it in `--devmode` and then confined it works, but indeed this is a blocker.

So, I'm moving to `classic`, while probably using just `devmode` would be fine. It all depends how we want this to interact with the system, and if the idea of having this confined is the final target.
In general using a `NextCloud` folder + symlinks would be enough to sync everything even when in devmode (and ultimately confined)...

So @enoch85 I guess we need your word and let us know how you prefer the snap to be.

**Comment by kyrofa** [comment]
> Ah, I've noticed that too when you only build it confined. If you run it in --devmode and then confined it works, but indeed this is a blocker.

Wait... what? Why? The example client snap to which I linked runs in confinement. What's missing here?

**Comment by 3v1n0** [comment]
@kyrofa oh, sorry that was because it was missing the `network-bind` interface... So, yeah it works confined.

So let's decide if for now we prefer it to be classic or devmode.

**Comment by enoch85** [comment]
@kyrofa Can you please have another look, this is more your area than mine. 

Thanks!

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [inline linux/snap/snapcraft.yaml:61]: Why is network-manager required?

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [inline linux/snap/snapcraft.yaml:32]: What is the purpose behind this? The commit mentions remembering config, but I don't have that issue with `nextcloud-client-kyrofa`. Am I misunderstanding?

File: linux/snap/snapcraft.yaml
Comment (3v1n0) [reply linux/snap/snapcraft.yaml:61]: See [this bug](https://bugs.launchpad.net/snappy/+bug/1650520), Qt network status checker needs it, and so this client if it wants to know the connection status.

This is not an auto-connected interface, though.

File: linux/snap/snapcraft.yaml
Comment (3v1n0) [reply linux/snap/snapcraft.yaml:32]: Without this, the client won't be able to access to `$SNAP/etc/Nextcloud/sync-exclude.lst` (which causes syncing not to work here; I deleted the log stating it, but if you want, I can paste it), the same with any other possibly hardcoded path. We need to be sure that the client will be able to access to such files.

Comment (enoch85) [reply linux/snap/snapcraft.yaml:32]: cc @kyrofa

**Comment by enoch85** [comment]
cc @kyrofa Can you please review?

**Comment by enoch85** [comment]
After discussing with @3v1n0 on IRC I think we should go for strict confinement because: `Snaps in developer mode can not be released into the stable and candidate store channels`. Though according to him there is a bug that makes the client ask for the user password every time the snap (client) starts which is not ideal. Maybe you could point out that bug so that more people could have a look?

Anyway, would be cool to get this merged asap. Waiting for what @kyrofa has to say.

**Comment by 3v1n0** [comment]
[This is the bug for the password saving](https://bugs.launchpad.net/snapd/+bug/1653769).

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [reply linux/snap/snapcraft.yaml:32]: Have you tried simply `-DSYSCONF_INSTALL_DIR=config` (or etc if you prefer, though that meaning breaks down in a snap)? That works for me.

**Comment by kyrofa** [comment]
> I think we should go for strict confinement because: `Snaps in developer mode can not be released into the stable and candidate store channels.` Though according to him there is a bug that makes the client ask for the user password every time the snap (client) starts which is not ideal.

This is exactly why I was suggesting classic confinement, which allows for no password prompt and release to stable channels at the expense of confinement. My only question is: what happens to the currently-installed snaps (using classic confinement) if that keyring interface lands and we move to strict confinement? Bad things?

If we're okay with the password prompt until that interface lands, then I'm all for strict confinement. But like I said, the prompt at startup will cause _me_ to prefer the deb version.

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [reply linux/snap/snapcraft.yaml:61]: Ah, okay. It still works if the interface is not connected, though? What features do we lose by not having it automatically connected?

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [reply linux/snap/snapcraft.yaml:44]: Fair enough, thank you :smiley: .

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [reply linux/snap/snapcraft.yaml]: Sounds good, thanks.

**Comment by 3v1n0** [comment]
> If we're okay with the password prompt until that interface lands, then I'm all for strict confinement. But like I said, the prompt at startup will cause me to prefer the deb version.

I agree with you @kyrofa, but at the same time releasing with classic interface would need more work for now (as the desktop-helper needs some love too, as some things such as `LD_LIBRARY_PATH` aren't set and there would no need to repack the system themes for example)...

Also, if we release with classic, we still need to set the `$HOME` to `$SNAP_USER_DATA` because otherwise moving to `confined` will cause the users the trouble to move their current synced dir to somewhere else without any notice.

Instead by doing this change now, we can just say this to all the users that will install the snap. So they will know it and nothing will change.

File: linux/snap/snapcraft.yaml
Comment (3v1n0) [reply linux/snap/snapcraft.yaml:61]: Things will work properly anyway... I'm not sure if this client is actually using this feature, but consider that as an "optimization". When Qt knows if there's a connection or not, it won't try to transfer data.

Users will be free to manually connect to this interface if they want an optimized client though.

**Comment by kyrofa** [comment]
> Instead by doing this change now, we can just say this to all the users that will install the snap. So they will know it and nothing will change.

Sounds good. I can't imagine that keyring interface will be that difficult, anyway.

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [reply linux/snap/snapcraft.yaml:61]: Good deal.

**Comment by 3v1n0** [comment]
> Sounds good. I can't imagine that interface will be that difficult, anyway.

Not really (and I'm looking into that), but... There still could be concerns as once the user has unlocked the keyring and if he saves the informations for the session, then I think that any app could access to such credentials (I've to check the APIs better though).

**Comment by kyrofa** [comment]
> There still could be concerns as once the user has unlocked the keyring and if he saves the informations for the session, then I think that any app could access to such credentials [...]

Indeed, though if that's true it'll likely be an interface that is not automatically connected, which still works. That's up to the security team, anyway :wink: .

File: linux/snap/snapcraft.yaml
Comment (3v1n0) [reply linux/snap/snapcraft.yaml:32]: I tried that, but then my snap was trying to access located in `$PATH_WHERE_I_WAS_BUILDING_THIS_SNAP/part/client/build/config` not sure why...

Comment (kyrofa) [reply linux/snap/snapcraft.yaml:32]: Do you get the same problem with `nextloud-client-kyrofa`?

Comment (kyrofa) [reply linux/snap/snapcraft.yaml:32]: This might be due to your setting CMAKE_INSTALL_PREFIX (why are you setting that?).

Comment (3v1n0) [reply linux/snap/snapcraft.yaml:32]: For the same reason, to make all the absolute paths to point to the right place... I'll try to get rid it and see... But as said there might be also other absolute paths here and there that wouldn't work otherwise.

Comment (3v1n0) [reply linux/snap/snapcraft.yaml:32]: Also translations are there... For example:

`for i in $(grep /snap/nextcloud-client prime/ -rl); do echo - $i:; strings $i | grep /snap/nextcloud-client; done` gives me:

```
- prime/usr/bin/nextcloud:
/snap/nextcloud-client/current/usr//snap/nextcloud-client/current/usr/lib/nextcloud
/snap/nextcloud-client/current/usr/share/nextcloud/i18n/
- prime/usr/bin/nextcloudcmd:
/snap/nextcloud-client/current/usr//snap/nextcloud-client/current/usr/lib/nextcloud
- prime/usr/lib/libnextcloudsync.so.2.3.0:
/snap/nextcloud-client/current/usr//snap/nextcloud-client/current/usr/lib/nextcloud
/snap/nextcloud-client/current/etc/%1
/snap/nextcloud-client/current/etc/%1/%1.conf
```

Let me see what happens using your way...

**Comment by enoch85** [comment]
So `strict` it is then?

What is left do do before this can be merged?

**Comment by enoch85** [comment]
@oparoz Can you please also take a look at this? You need to register the name under the Nextcloud store account, and afaik you are the only one with access.

I also suggest that we release this in the same phase as the ordinary ownCloud client, but without betas.

**Comment by kyrofa** [comment]
@3v1n0 how do you envision this being released? Since the `snapcraft.yaml` is hidden away, I don't think Launchpad will build this.

**Comment by 3v1n0** [comment]
@kyrofa check this branch: https://github.com/3v1n0/nextcloud-client_theming/blob/travis-snap-build/linux/travis-build.sh (and the relative `.travis.yaml`)

Also I think we'd need a way to tell launchpad where the snapcraft.yaml can be found. As it's not always in `snap` or root folder.

**Comment by 3v1n0** [comment]
@enoch85 `strict` is fine for me, if we really have troubles with it, we can go with something more permissive...

As for release, the branch linked above, will release any git revision to `edge`. As for manual releases, to stable, it's up to you guys to define the rules.

PS: @oparoz you can add multiple accounts to manage a snap, in the store collaboration section.

File: linux/snap/snapcraft.yaml
Comment (3v1n0) [reply linux/snap/snapcraft.yaml:32]: I've tried that too, but it's not fine, the same command (with different pattern) gives me:

```
/media/marco/M.2-SSD/client_theming/linux/parts/client/build/client/config/%1
/media/marco/M.2-SSD/client_theming/linux/parts/client/build/client/config/%1/%1.conf
```

where `/media/marco/M.2-SSD/client_theming` is the local repo path.

Comment (3v1n0) [reply linux/snap/snapcraft.yaml:32]: I've tried that too, but it's not fine, the same command (with different pattern) gives me:

```
/media/marco/M.2-SSD/client_theming/linux/parts/client/build/client/config/%1
/media/marco/M.2-SSD/client_theming/linux/parts/client/build/client/config/%1/%1.conf
```

where `/media/marco/M.2-SSD/client_theming` is the local repo path.

**Comment by 3v1n0** [comment]
If you want to test a built snap in this way, here's a binary: https://transfer.sh/XW1As/nextcloud-client-2.2.4-git-amd64.snap

**Comment by oparoz** [comment]
@3v1n0 We can make it so that Nextcloud is the publisher of the snaps you create.

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [reply linux/snap/snapcraft.yaml:44]: I just saw [this](https://github.com/snapcore/snapd/pull/2765), by the way. Can you use the one in /usr/local/bin of the core snap?

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [reply linux/snap/snapcraft.yaml:32]: You say you get a log denying access. Do you still get that message? Do you get that message with nextcloud-client-kyrofa?

File: linux/snap/snapcraft.yaml
Comment (kyrofa) [reply linux/snap/snapcraft.yaml:44]: No scratch that, it's not available until 2.23. Carry on.

**Comment by 3v1n0** [comment]
@oparoz of course... That's the plan!
See [this page](https://snapcraft.io/docs/build-snaps/publish), you need to [register the snap name](https://myapps.developer.ubuntu.com/dev/click-apps/register-name/) (I chose `nextcloud-client` but you can change it if you want), and then we should make travis to upload the snap on each git release (I've already a [branch](https://github.com/3v1n0/nextcloud-client_theming/blob/travis-snap-build/linux/travis-build.sh) for that) to the edge channel.

But all the snap concept is around the fact that you (upstream) will then manage the releases...

No others will be able to upload it then.

File: linux/snap/snapcraft.yaml
Comment (3v1n0) [reply linux/snap/snapcraft.yaml:32]: Mh, no... I can't either test what's the log saying though since the `run-client`is missing a `$*`.
I'll try to build it locally

File: linux/snap/snapcraft.yaml
Comment (3v1n0) [reply linux/snap/snapcraft.yaml:44]: Eh, in fact that's why I didn't do it :-)

**Comment by oparoz** [comment]
I registered `nextcloud-client`. Travis can't build the snap, we need to trigger builds on Launchpad AFAIK.

**Comment by 3v1n0** [comment]
@oparoz good.. 

> Travis can't build the snap, we need to trigger builds on Launchpad AFAIK. 

Of course it does: https://travis-ci.org/3v1n0/nextcloud-client_theming/builds/204436924

**Comment by oparoz** [comment]
Ah, yes, we don't run into the same timeout issues we have with the nextcloud server snap.

So what are the next steps then?

1. Merge this
1. Set up Travis script
1. Manually upload the first build
1. Give you access as a collaborator
1. Automate the uploading

?

**Comment by kyrofa** [comment]
Will you be able to build for both amd64 and i386 on Travis?

**Comment by 3v1n0** [comment]
@kyrofa of course, it's already doing it, see the [travis matrix](https://github.com/3v1n0/nextcloud-client_theming/blob/travis-snap-build/.travis.yml#L15).

**Comment by 3v1n0** [comment]
@oparoz 

Once we decided if you want to go with devmode or strict confinement (I guess you accept for the strict limitations for now) the steps are

1. Merge this (and generate the snaps you want to upload manually)
2. Wait for my new PR that adds the travis building
3. Generate the .snapcraft.cfg file and save the credentials at travis level to make travis to automatically upload the file (you can do it with `snapcraft enable-ci travis`)

I don't need to be a collaborator for the snap, as you can manage that alone, but I can help if you want...

PS: @kyrofa please re-consider [your review](https://github.com/nextcloud/client_theming/pull/77#pullrequestreview-22663513) as that issue seems to be addressed.

**Comment by kyrofa** [comment]
> @kyrofa please re-consider your review as that issue seems to be addressed.

It has not. You still haven't answered [my question](https://github.com/nextcloud/client_theming/pull/77#discussion_r104180119).

**Comment by 3v1n0** [comment]
@kyrofa which one you mean, am I missing something? The one related to `xdg-open` seems replied to me, or what do you miss?

**Comment by kyrofa** [comment]
> @kyrofa which one you mean, am I missing something?

Sorry, I added [the link](https://github.com/nextcloud/client_theming/pull/77#discussion_r104180119) in an edit. I don't like the absolute path hack, and the spike I built seemed to work without it. You imply that there are issues, but you haven't addressed why those issues don't seem to be present in my spike. Perhaps they _are_ present and I'm not seeing them for some reason, but you haven't addressed that either.

**Comment by 3v1n0** [comment]
@kyrofa ah, ok... Since the review was pointing to something else.
Let me try building your snap, however as I said during compilation there are some absolute paths that are hardcoded, and we have to ensure that these will point to the right place.

**Comment by kyrofa** [comment]
> @kyrofa ah, ok... Since the review was pointing to something else.

Yeah I don't typically "request changes" multiple times. The red X is already there :stuck_out_tongue: .

> Let me try building your snap, however as I said during compilation there are some absolute paths that are hardcoded, and we have to ensure that these will point to the right place.

Or you can just install from edge as I've suggested. The snapcraft.yaml will pull a tag that doesn't build without modifications (i.e. it was pre-[this-PR](https://github.com/owncloud/client/pull/5401)).

**Comment by 3v1n0** [comment]
Yeah @kyrofa, I already installed it from edge, but as said, I wasn't able to see the full log since you can't pass arguments to that version... So I wanted to rebuild it, however even the snap in the store points to the wrong conf path:

```
/home/ubuntu/nextcloud-client-snap/parts/client/build/client/config/%1
/home/ubuntu/nextcloud-client-snap/parts/client/build/client/config/%1/%1.conf
```

Building is the same, of course [as for the GIT_SHA edit, I already figured that, but thanks :-)]...

PS: sorry I've just noticed I didn't submit some comments

**Comment by 3v1n0** [comment]
@kyrofa so I've just rebuilt your snap... It works when using version (git tag) 2.2.4, not with master:

See this log: http://paste.ubuntu.com/24127025/ :
`03-06 22:32:55:921 2679 OCC::Folder::setIgnoredFiles: ==== adding system ignore list to csync: "/media/marco/M.2-SSD/nextcloud-client-snap/parts/client/build/client/config/Nextcloud/sync-exclude.lst"`

And the client error:

image

**Comment by kyrofa** [comment]
> @kyrofa so I've just rebuilt your snap... It works when using version (git tag) 2.2.4, not with master:

Ah ha! That's what I wanted to know! That explains the discrepancy. I wonder what changed.

With that in mind, I'm okay with this, but I very very much dislike the hack. We should consider patches to make the client more flexible, because things like this can get in the way of cross-distro support. For example, this is currently some output from a Fedora machine:

```
$ hello-world.env | grep SNAP
SNAP_USER_COMMON=/home/<user>/snap/hello-world/common
SNAP_REEXEC=
SNAP_LIBRARY_PATH=/var/lib/snapd/lib/gl:
SNAP_COMMON=/var/snap/hello-world/common
SNAP_USER_DATA=/home/<user>/snap/hello-world/27
SNAP_DATA=/var/snap/hello-world/27
SNAP_REVISION=27
SNAP_NAME=hello-world
SNAP_ARCH=amd64
SNAP_VERSION=6.3
SNAP=/var/lib/snapd/snap/hello-world/27
```

Note how `$SNAP` is in a different place. This will hopefully change soon, but it's something to keep in mind.

**Review by kyrofa — APPROVED** [review]
I approve with the caveat that we should investigate making the client more flexible at runtime. Using absolute paths and an organize is a hack from which we should move away as soon as we can.

**Comment by 3v1n0** [comment]
Ouch, I though this was somewhat decided that `/snap/$SNAPCRAFT_PROJECT_NAME/current` was always pointing to the snap current location... Is that because on fedora we've still an old snapd?

Keep in mind that most of desktop apps we snapped already use this assertion... 

We can probably fix things in code, but it's annoying for now, as I think it needs some deeper changes.

**Comment by kyrofa** [comment]
> Ouch, I though this was somewhat decided that /snap/$SNAPCRAFT_PROJECT_NAME/current was always pointing to the snap current location... Is that because on fedora we've still an old snapd?

An older snapd and an older snap-confine. Other distros are the same. Things trickle down, I'm afraid. I see code in the newest snap-confine that mentions dealing with this issue, but I have yet to see it in use.

> Keep in mind that most of desktop apps we snapped already use this assertion...

Any time someone asked me this I said it was a bad idea for this very reason.

**Comment by enoch85** [comment]
Hope I wasn't too quick on the merge button...

**Comment by 3v1n0** [comment]
I think this is fine for now, it's just an ubuntu-only solution, but we can improve in different branch.

**Comment by oparoz** [comment]
Could someone track @kyrofa's review comment in an issue so that we don't forget to address it as soon as we can?

**Pull request merged**
